File: api-gateway-java-reactive/src/main/java/ru/kuramshindev/apigatewayjavareactive/GatewayController.java
package ru.kuramshindev.apigatewayjavareactive;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;
import reactor.core.publisher.Mono;
import reactor.util.function.Tuple2;
import ru.kuramshin.dto.apigateway.ApiGatewayDto;

import java.util.UUID;

@Slf4j
@RequiredArgsConstructor
@RestController("gateway-controller")
@RequestMapping(value = "/api/v1")
public class GatewayController {

    private final GatewayService gatewayService;

    @RequestMapping(method = RequestMethod.GET, value = "/contracts/{uuid}")
    @ResponseStatus(HttpStatus.OK)
    public Mono<ApiGatewayDto> getContractById(@PathVariable("uuid") UUID uuid) {
        return gatewayService.getContractById(uuid)
                .elapsed()
                .doOnNext(tuple -> log.info("Elapsed time: {}", tuple.getT1()))
                .map(Tuple2::getT2);
    }
}

File: api-gateway-kotlin-reactive/src/main/kotlin/ru/kuramshindev/apigatewaykotlinreactive/GatewayController.kt
package ru.kuramshindev.apigatewaykotlinreactive

import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*
import ru.kuramshin.dto.apigateway.ApiGatewayDto
import java.util.*
import kotlin.system.measureTimeMillis

@RestController("gateway-controller")
@RequestMapping(value = ["/api/v1"])
class GatewayController(
    private val gatewayService: GatewayService
) {

    private val log = LoggerFactory.getLogger(javaClass)

    @RequestMapping(method = [RequestMethod.GET], value = ["/contracts/{uuid}"])
    @ResponseStatus(HttpStatus.OK)
    suspend fun getContractById(@PathVariable("uuid") uuid: UUID): ApiGatewayDto {
        var apiGatewayDto: ApiGatewayDto
        val elapsed = measureTimeMillis {
            apiGatewayDto = gatewayService.getContractById(uuid)
        }
        log.info("Elapsed time: $elapsed")
        return apiGatewayDto
    }
}